src/Negotiation/Negotiator.php

```diff
@@ -21,23 +21,62 @@ protected function match(AcceptHeader $accept, AcceptHeader $priority, $index)
             return null;
         }
 
-        $ab = $accept->getBasePart();
-        $pb = $priority->getBasePart();
+        $acceptBase = $accept->getBasePart();
+        $priorityBase = $priority->getBasePart();
 
-        $as = $accept->getSubPart();
-        $ps = $priority->getSubPart();
+        $acceptSub = $accept->getSubPart();
+        $prioritySub = $priority->getSubPart();
 
         $intersection = array_intersect_assoc($accept->getParameters(), $priority->getParameters());
 
-        $baseEqual = !strcasecmp($ab, $pb);
-        $subEqual  = !strcasecmp($as, $ps);
+        $baseEqual = !strcasecmp($acceptBase, $priorityBase);
+        $subEqual  = !strcasecmp($acceptSub, $prioritySub);
 
-        if (($ab === '*' || $baseEqual) && ($as === '*' || $subEqual) && count($intersection) === count($accept->getParameters())) {
+        if (($acceptBase === '*' || $baseEqual)
+            && ($acceptSub === '*' || $subEqual)
+            && count($intersection) === count($accept->getParameters())
+        ) {
             $score = 100 * $baseEqual + 10 * $subEqual + count($intersection);
 
             return new Match($accept->getQuality() * $priority->getQuality(), $score, $index);
         }
 
+        if (!strstr($acceptSub, '+') || !strstr($prioritySub, '+')) {
+            return null;
+        }
```

Comment:
If either the accept header or the priority do not have `+` segments, we can return immediately without further checks, retaining the previous workflow.